// Copyright 2026 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


syntax = "proto3";

package ray.rpc.events;

import "google/protobuf/timestamp.proto";

// Message containing the execution information of a submission job.
// It can be used to capture the full state transition history.
//
// For static information associated with this event, see SubmissionJobDefinitionEvent.
message SubmissionJobLifecycleEvent {
  enum State {
    UNSPECIFIED = 0;
    // The job has not started yet, likely waiting for the runtime_env to be set up.
    PENDING = 1;
    // The job is currently running.
    RUNNING = 2;
    // The job was intentionally stopped by the user.
    STOPPED = 3;
    // The job finished successfully.
    SUCCEEDED = 4;
    // The job failed.
    FAILED = 5;
  }

  message StateTransition {
    // The state of the job after this transition.
    State state = 1;
    // Timestamp when the transition occurred.
    google.protobuf.Timestamp timestamp = 2;
    // A message describing the status in more detail.
    string message = 3;
    // Error type (e.g. Runtime env setup failure, Internal error, user script error).
    string error_type = 4;
    // The node id that driver is running on.
    bytes driver_node_id = 5;
    // Driver agent http address.
    string driver_agent_http_address = 6;
    // The driver process exit code after the driver executed.
    // Note: 0 is a valid exit code, use has_driver_exit_code() to check if set.
    optional int32 driver_exit_code = 7;
  }

  // Unique submission job identifier.
  string submission_id = 1;
  // This records the state transitions within each export interval. The consumer should
  // concatenate these intervals over the job's lifetime to reconstruct the complete
  // state transition time series.
  repeated StateTransition state_transitions = 2;
}
